### ISSUE TITLE
Code for hyperparameter optimization of model

### ISSUE THREAD

**Issue by abuzarmahmood** [pulakatlab/auto_slide#13]

**Comment by abuzarmahmood** [comment]
The issue involves implementing hyperparameter optimization for a model in the `pulakatlab/auto_slide` repository. The relevant files identified for modification are:

1. **`prediction.py`**: Handles model predictions and may need adjustments to incorporate optimized hyperparameters during prediction.
2. **`training.py`**: Responsible for training the model and is a key candidate for implementing hyperparameter optimization techniques.
3. **`run_pipeline.py`**: Orchestrates the entire pipeline, including training and prediction, and may need modifications to integrate hyperparameter optimization.

### Plan to Address the Issues
1. Implement hyperparameter optimization in the training process.
2. Ensure that the optimized hyperparameters are used during prediction.
3. Integrate these changes into the pipeline.

### Specific Changes Suggested

1. **Implement Hyperparameter Optimization in Training**
   - **File:** `src/pipeline/model/training.py`
   - **Description of Change:** Use a library like `Optuna` or `GridSearchCV` to perform hyperparameter optimization during the training process.
   - **Code Snippet:**
     ```python
     import optuna

     def objective(trial):
         # Define hyperparameters to optimize
         lr = trial.suggest_loguniform('lr', 1e-5, 1e-1)
         momentum = trial.suggest_uniform('momentum', 0.5, 0.9)
         
         # Update optimizer with trial hyperparameters
         optimizer = torch.optim.SGD(params, lr=lr, momentum=momentum, weight_decay=0.0005)
         
         # Training loop with trial parameters
         for epoch in range(n_epochs):
             # Training logic here
             pass
         
         # Return validation loss as objective value
         return val_epoch_loss

     study = optuna.create_study(direction='minimize')
     study.optimize(objective, n_trials=100)
     best_params = study.best_params
     ```

2. **Use Optimized Hyperparameters in Prediction**
   - **File:** `src/pipeline/model/prediction.py`
   - **Description of Change:** Ensure that the model uses the best hyperparameters found during training.
   - **Code Snippet:**
     ```python
     # Load best hyperparameters
     best_lr = best_params['lr']
     best_momentum = best_params['momentum']
     
     # Use best hyperparameters for prediction
     optimizer = torch.optim.SGD(params, lr=best_lr, momentum=best_momentum, weight_decay=0.0005)
     ```

3. **Integrate Changes into the Pipeline**
   - **File:** `src/pipeline/run_pipeline.py`
   - **Description of Change:** Modify the pipeline to incorporate hyperparameter optimization.
   - **Code Snippet:**
     ```python
     def run_model_training(data_dir=None):
         logger.info("Starting model training with hyperparameter optimization...")
         from model.training import main as training_main
         
         try:
             training_main()  # Ensure this function uses the new optimization logic
             logger.info("Model training completed successfully with optimized hyperparameters")
             return True
         except Exception as e:
             logger.error(f"Error in model training: {str(e)}")
             return False
     ```

These changes aim to implement hyperparameter optimization in the model training process and ensure that the optimized parameters are used during prediction, thereby improving the model's performance.

---
*This response was automatically generated by blech_bot using model gpt-4o*

---
*This response was automatically generated by blech_bot*

**Comment by abuzarmahmood** [comment]
Based on the user feedback and the content of the files, here is an improved response that incorporates the intersection over union (IoU) metric for hyperparameter evaluation:

### Plan to Address the Issues
1. Implement hyperparameter optimization in the training process using IoU as the evaluation metric.
2. Ensure that the optimized hyperparameters are used during prediction.
3. Integrate these changes into the pipeline.

### Specific Changes Suggested

1. **Implement Hyperparameter Optimization in Training**
   - **File:** `src/pipeline/model/training.py`
   - **Description of Change:** Use a library like `Optuna` to perform hyperparameter optimization during the training process, evaluating the model based on the IoU metric.
   - **Code Snippet:**
     ```python
     import optuna
     from sklearn.metrics import jaccard_score

     def objective(trial):
         # Define hyperparameters to optimize
         lr = trial.suggest_loguniform('lr', 1e-5, 1e-1)
         momentum = trial.suggest_uniform('momentum', 0.5, 0.9)
         
         # Update optimizer with trial hyperparameters
         optimizer = torch.optim.SGD(params, lr=lr, momentum=momentum, weight_decay=0.0005)
         
         # Training loop with trial parameters
         for epoch in range(n_epochs):
             # Training logic here
             pass
         
         # Calculate IoU for validation set
         iou_scores = []
         with torch.no_grad():
             for imgs, targets in val_dl:
                 preds = model(imgs)
                 for pred, target in zip(preds, targets):
                     iou = jaccard_score(target['masks'].cpu().numpy().flatten(), pred['masks'].cpu().numpy().flatten(), average='macro')
                     iou_scores.append(iou)
         
         # Return negative IoU as objective value (since Optuna minimizes)
         return -np.mean(iou_scores)

     study = optuna.create_study(direction='maximize')
     study.optimize(objective, n_trials=100)
     best_params = study.best_params
     ```

2. **Use Optimized Hyperparameters in Prediction**
   - **File:** `src/pipeline/model/prediction.py`
   - **Description of Change:** Ensure that the model uses the best hyperparameters found during training.
   - **Code Snippet:**
     ```python
     # Load best hyperparameters
     best_lr = best_params['lr']
     best_momentum = best_params['momentum']
     
     # Use best hyperparameters for prediction
     optimizer = torch.optim.SGD(params, lr=best_lr, momentum=best_momentum, weight_decay=0.0005)
     ```

3. **Integrate Changes into the Pipeline**
   - **File:** `src/pipeline/run_pipeline.py`
   - **Description of Change:** Modify the pipeline to incorporate hyperparameter optimization.
   - **Code Snippet:**
     ```python
     def run_model_training(data_dir=None):
         logger.info("Starting model training with hyperparameter optimization...")
         from model.training import main as training_main
         
         try:
             training_main()  # Ensure this function uses the new optimization logic
             logger.info("Model training completed successfully with optimized hyperparameters")
             return True
         except Exception as e:
             logger.error(f"Error in model training: {str(e)}")
             return False
     ```

These changes aim to implement hyperparameter optimization using IoU as the evaluation metric in the model training process and ensure that the optimized parameters are used during prediction, thereby improving the model's performance.



---
*This response was automatically generated by blech_bot using model gpt-4o*

---
*This response was automatically generated by blech_bot*

**Comment by abuzarmahmood** [comment]
Recording the current hyperparameters as the baseline for any search, tracked
under #116.

From `setup_training` and `train_model`:

| Parameter | Value | Note |
| --- | --- | --- |
| optimizer | SGD | |
| lr | 0.005 | **constant for all 90 epochs** |
| momentum | 0.9 | |
| weight_decay | 0.0005 | |
| n_epochs | 90 | |
| batch_size | 2 | structurally fixed, see #122 |
| lr schedule | none | |
| warmup | none | |
| grad clipping | none | see #119 |

The absence of a schedule is the notable one. `lr=0.005` held flat for 90 epochs
means the run never anneals, so the final weights sit wherever the last few
noisy batch-size-2 gradients left them — and the checkpoint is selected on a
validation loss that is itself noisy (#118).

Before spending compute on a search, the cheap structural wins are probably:

```python
warmup = torch.optim.lr_scheduler.LinearLR(
    optimizer, start_factor=0.001, total_iters=500)
main = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=n_epochs)
```

or `MultiStepLR(optimizer, milestones=[60, 80], gamma=0.1)`, which is the
torchvision detection reference recipe.

Also worth noting that `lr` and `batch_size` should be searched jointly rather
than independently, and that #122 has to land before `batch_size` is searchable
at all.
